src/logic/map_objects/checkstep.cc

```diff
@@ -121,7 +121,12 @@ CheckStepFerry
 */
 bool CheckStepFerry::allowed(
    const Map& map, const FCoords& from, const FCoords& to, int32_t dir, CheckStep::StepId) const {
+	if (!(to.field->nodecaps() & MOVECAPS_WALK) && !(to.field->nodecaps() & MOVECAPS_SWIM)) {
+		// can't swim on lava
+		return false;
+	}
 	if (MOVECAPS_SWIM & (from.field->nodecaps() | to.field->nodecaps())) {
```

Comment:
Do you need to check the `to.field->nodecaps()` here again?

Reply:
The first `if` checks whether the destination field is near lava, in which case we know immediately that we mustn't go there. The second `if` checks whether either field is open sea, in which case we know the move is allowed. When we get to this `if`, we know that `to.field` has `MOVECAPS_WALK` *or* `MOVECAPS_SWIM`, but we still need to check which one.
Often, both fields will have `MOVECAPS_WALK` but neither has `MOVECAPS_SWIM` so we need to check at large with the `switch` below.

Reply:
Now I see it. Thanks for the explanation :)